plugins/source_testing.go

```diff
@@ -23,34 +24,52 @@ func init() {
 	_ = faker.SetRandomMapAndSliceMaxSize(1)
 }
 
-// type
-
 func TestSourcePluginSync(t *testing.T, plugin *SourcePlugin, spec specs.Source) {
-	// t.Parallel()
 	t.Helper()
-	// No need for configuration or db connection, get it out of the way first
-	// testTableIdentifiersForProvider(t, resource.Provider)
 
-	// l := testlog.New(t)
-	// l.SetLevel(hclog.Info)
-	// resource.Plugin.Logger = l
-	resources := make(chan *schema.Resource)
+	resourcesChannel := make(chan *schema.Resource)
 	var fetchErr error
 
 	go func() {
-		defer close(resources)
-		fetchErr = plugin.Sync(context.Background(), spec, resources)
+		defer close(resourcesChannel)
+		fetchErr = plugin.Sync(context.Background(), spec, resourcesChannel)
 	}()
-	totalResources := 0
-	for resource := range resources {
-		totalResources++
-		validateResource(t, resource)
+
+	syncedResources := make([]*schema.Resource, 0)
+	for resource := range resourcesChannel {
+		syncedResources = append(syncedResources, resource)
 	}
-	if fetchErr != nil {
-		t.Fatal(fetchErr)
+	require.NoError(t, fetchErr)
+
+	validateTables(t, plugin.Tables(), syncedResources)
+}
+
+func getTableResource(t *testing.T, table *schema.Table, resources []*schema.Resource) *schema.Resource {
+	t.Helper()
+	for _, resource := range resources {
+		if resource.Table.Name == table.Name {
+			return resource
+		}
 	}
-	if totalResources == 0 {
-		t.Fatal("no resources fetched")
+
+	return nil
+}
+
+func validateTable(t *testing.T, table *schema.Table, resources []*schema.Resource) {
+	t.Helper()
+	resource := getTableResource(t, table, resources)
+	if resource != nil {
+		validateResource(t, resource)
+	} else {
+		t.Errorf("Expected table %s to be synced but it was not found", table.Name)
+	}
```

Comment:
Bit of a nit, but typically linters will tell you to remove the `else` here and change it to something like:

```
if resource == nil {
   t.Errorf("Expected table %s to be synced but it was not found", table.Name)
   return
}
validateResource(t, resource)
```

This way it's a bit easier to follow the happy path logic by reading the top-level indent, with error handling happening on deeper indents.

Reply:
Good catch, fixed in https://github.com/cloudquery/plugin-sdk/pull/85/commits/c3f09268fa3065e0e1ea36736bd9e74d045084a2

Reply:
Thanks! It's really not that important in this specific case, but to align on a general style guideline, in Go I believe it's preferable to have the error condition inside the `if` condition, with the happy path outside (with least indent). So I would write:

```
if resource == nil {
   t.Errorf("Expected table %s to be synced but it was not found", table.Name)
   return
}
validateResource(t, resource)
```

rather than

```
if resource != nil {
   validateResource(t, resource)
   return
}
t.Errorf("Expected table %s to be synced but it was not found", table.Name)
```

This way, when reading a function, the reader's eye can stay on the left-hand side and quickly see what a function does without needing to read about how error conditions are handled, unless they are interested in that aspect. Bonus points in this case for have a simple equality operator (`==`) over a negated one (`!=`) in the `if` condition.

The authors and main contributors of Go have discussed these points in various documents, but I'd be hard-pressed to find them now 😄

Reply:
Cool, thanks for the additional context, I missed it in your first comment. Please let me know if https://github.com/cloudquery/plugin-sdk/pull/85/commits/78be2b3e71bfc4b2439a6b0f024227ce7075f545 works

Reply:
Thanks, looks good! Didn't mean to nitpick, I've just found this a good rule of thumb for writing clean Go code, so I hope it's at least useful advice :)

Reply:
💯 Agree. I looked for a linter to enforce this but couldn't find one

Reply:
I've been looking for an excuse to write my own linter! 😄

Reply:
I'm pretty sure I'm not hallucinating that I've been warned about similar kinds of things by tooling in the past (especially a function that ends in if/else where the if branch has a return), but I also cannot find it now.

Reply:
I think we have https://github.com/mgechev/revive/blob/master/RULES_DESCRIPTIONS.md#early-return. Maybe it doesn't strictly fall under that rule